### PR TITLE
dmc cartpole

### PR DESCRIPTION
**Pull request by Benjamin-eecs**
#103

### CODE REVIEW

File: envpool/mujoco/dmc/cartpole.h [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+#endif
+  }
+
+  mjtNum CartPosition() {
+    // return self.named.data.qpos['slider'][0]
+    return data_->qpos[id_slider_];
+  }
+  std::array<mjtNum, 3> AngularVel() {
+    // return self.data.qvel[1:]
+    std::array<mjtNum, 3> angular_vel;
```
Comment (Benjamin-eecs) [inline envpool/mujoco/dmc/cartpole.h:207]: any elegant solution dealing with changing of pole numbers?

File: envpool/mujoco/dmc/cartpole.h [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+      angular_vel[i] = data_->qvel[1 + i];
+    }
+    for (int i = n_poles_; i < 3; ++i) {
+      angular_vel[i] = std::numeric_limits<double>::infinity();
+    }
+    return angular_vel;
+  }
+  std::array<mjtNum, 3_> PoleAngleCosine() {
+    // return self.named.data.xmat[2:, 'zz']
+    std::array<mjtNum, 3> pole_angle_cosine;
```
Comment (Benjamin-eecs) [inline envpool/mujoco/dmc/cartpole.h:218]: any elegant solution dealing with changing of pole numbers?

File: envpool/mujoco/dmc/cartpole.h [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    }
+    for (int i = n_poles_; i < 3; ++i) {
+      pole_angle_cosine[i] = std::numeric_limits<double>::infinity();
+    }
+    return pole_angle_cosine;
+  }
+  std::array<mjtNum, 7> BoundedPosition() {
+    // return np.hstack((self.cart_position(),
+    //                   self.named.data.xmat[2:, ['zz', 'xz']].ravel()))
+    std::array<mjtNum, 7> bounded_position;
```
Comment (Benjamin-eecs) [inline envpool/mujoco/dmc/cartpole.h:230]: any elegant solution dealing with changing of pole numbers?

File: envpool/mujoco/dmc/utils.cc [hunk 4]
```diff
@@ -55,6 +55,58 @@ std::string XMLRemoveByBodyName(const std::string& content,
   return writer.result;
 }
 
+std::string XMLAddPoles(const std::string& content, int pole_numbers) {
+  pugi::xml_document doc;
```
Comment (Benjamin-eecs) [inline envpool/mujoco/dmc/utils.cc:59]: due to time limit, needed help here.

File: envpool/make_test.py [hunk 5]
```diff
@@ -130,6 +130,10 @@ def test_make_mujoco_dmc(self) -> None:
         "AcrobotSwingup-v1",
         "AcrobotSwingupSparse-v1",
         "BallInCupCatch-v1",
+        "CartpoleBalance-v1",
+        "CartpoleBalanceSparse-v1",
+        "CartpoleSwingup-v1",
+        "CartpoleSwingupSparse-v1",
```
Comment (Benjamin-eecs) [inline envpool/make_test.py:136]: have no idea why make_test fail.

**Comment by Trinkle23897** [comment]
2022-05-22 09-44-23 的屏幕截图

**Comment by Trinkle23897** [comment]
```c++
#include <bits/stdc++.h>

#include <mjxmacro.h>
#include <mujoco.h>

#include "pugixml.hpp"

std::string GetFileContent(const std::string& base_path,
                           const std::string& asset_name) {
  // hardcode path here :(
  std::string filename = base_path + "/" + asset_name;
  std::ifstream ifs(filename);
  std::stringstream ss;
  ss << ifs.rdbuf();
  return ss.str();
}

struct xml_string_writer: pugi::xml_writer
{
    std::string result;

    virtual void write(const void* data, size_t size)
    {
        result.append(static_cast<const char*>(data), size);
    }
};

std::string GetManipulatorXML(const std::string &base_path,
                              const std::string &task_name)
{
    int n = 1;
    if (task_name == "two_poles") {
        n = 2;
    } else if (task_name == "three_poles") {
        n = 3;
    }
    std::string content = GetFileContent(base_path, "cartpole.xml");
    if (n == 1) {
        return content;
    }
    pugi::xml_document doc;
    doc.load_string(content.c_str());

    pugi::xml_node body = doc.select_node("//worldbody/body/body").node();
    for (int i = 2; i <= n; ++i) {
        pugi::xml_node new_pole = body.append_child("body");
        new_pole.append_attribute("childclass") = "pole";
        new_pole.append_attribute("name") = ("pole_" + std::to_string(i)).c_str();
        new_pole.append_attribute("pos") = "0 0 1";
        pugi::xml_node joint = new_pole.append_child("joint");
        joint.append_attribute("name") = ("hinge_" + std::to_string(i)).c_str();
        pugi::xml_node geom = new_pole.append_child("geom");
        geom.append_attribute("name") = ("pole_" + std::to_string(i)).c_str();
        body = new_pole;
    }
    
    pugi::xml_node floor = doc.select_node("//worldbody/geom").node();
    floor.attribute("pos").set_value(("0 0 " + std::to_string(1 - n - 0.05)).c_str());
    pugi::xpath_node_set cameras = doc.select_nodes("//worldbody/camera");
    for (const pugi::xpath_node &c : cameras) {
        std::string name = c.node().attribute("name").value();
        if (name == "fixed") {
            c.node().attribute("pos").set_value(("0 " + std::to_string(-1 - 2 * n) + " 1").c_str());
        } else if (name == "lookatcart") {
            c.node().attribute("pos").set_value(("0 " + std::to_string(-2 * n) + " 2").c_str());
        }
    }
    xml_string_writer writer;
    doc.print(writer);
    return writer.result;
}

int main(int argc, char **argv) {
    auto raw_xml = GetManipulatorXML("dmc", argv[1]);
    std::cout << raw_xml << std::endl;
}
```

File: envpool/make_test.py (hunk 5 above)
Comment (Trinkle23897) [reply envpool/make_test.py:136]: use `--config=debug`:
```bash
$ bazel test --test_output=all //envpool/mujoco:mujoco_dmc_align_test --config=debug --spawn_strategy=local
[ RUN      ] _MujocoDmcAlignTest.test_cartpole
I0522 14:58:34.664073 140272067917632 mujoco_dmc_align_test.py:153] align check for cartpole balance
WARNING: Logging before InitGoogleLogging() is written to STDERR
F20220522 14:58:34.665021 2979401 array.h:160] Check failed: sz == size (7 vs. 3)  assignment size mismatch
*** Check failure stack trace: ***
Fatal Python error: Aborted
```

**Review by Trinkle23897 — APPROVED** [review]

**Pull request merged**